### ISSUE TITLE
v8: Prop type warning during server-side rendering

### ISSUE THREAD

**Issue by zmdavis** [downshift-js/downshift#1525]
- `downshift` version: 8.0.0
- `node` version: 18.11.0
- `npm` (or `yarn`) version: npm 9.7.2

**What happened**:

In v8.0.0, the following warning is regularly printed to the console during server-side rendering:

```
Warning: Failed prop type: The prop `environment.addEventListener` is marked as required in `Downshift`, but its value is `undefined`.
```

**Problem description**:

I believe that due to [prop type changes](https://github.com/downshift-js/downshift/pull/1482) in v8, the prop types are inconsistent with the default value used for `environment` when server-side rendering (or in any environment where `window` is undefined).

The [default value](https://github.com/downshift-js/downshift/blob/master/src/downshift.js#L112) of `environment` is set to an empty object (`{}`) in this case, but an empty object violates the [shape defined](https://github.com/downshift-js/downshift/blob/master/src/downshift.js#L54-L64) for `environment`. This in turn triggers a `console.warn`

**Suggested solution**:

Either the default value of `environment` when `window` is undefined should be updated to be valid per the prop types, or the prop types should be updated to allow for optional properties.

I'm happy to put up a PR if I can get some guidance on which direction you'd prefer.

**Comment by silviuaavram** [comment]
I think we should remove the `isRequired` part from all the `environment` prop types, what do you think?

The original changes are in [this PR](https://github.com/downshift-js/downshift/pull/1463). Since the shape of the `environment` prop was correctly fixed in that PR, all we need is to remove the isRequired from all occurrences.

Let me know if you would like to contribute. Thank you for reporting!

**Comment by silviuaavram** [comment]
Actually I am still thinking about this issue.

According to the documentation, `environment`:
```
This prop is only useful if you're rendering downshift within a different window context from where your JavaScript is running; for example, an iframe or a shadow-root. If the given context is lacking document and/or add|removeEventListener on its prototype (as is the case for a shadow-root) then you will need to pass in a custom object that is able to provide [access to these properties](https://gist.github.com/Rendez/1dd55882e9b850dd3990feefc9d6e177) for downshift.
```

On SSR, since we don't have a `window` or equivalent, just don't pass the `environment` prop, as you cannot replace it with anything in your case.

If you indeed have an equivalent, then that equivalent must have all the properties marked as `required`. It's consistent with our typescript definition.

**Comment by zmdavis** [comment]
Thanks for following up @silviuaavram. I'm not sure if you've seen my [pull request](https://github.com/downshift-js/downshift/pull/1530) but I think you and I came to similar conclusions, though there's still some room for debate about the right fix.

My fix is to allow for either an `environment` of the proper shape or an empty object, since that's what the default is. With those changes someone could theoretically still pass in an empty object, which would be equivalent to passing in nothing. But if they specify any properties for `environment` they would need to specify all of the required properties.

There's probably a more robust fix that doesn't allow for passing in an empty object (which doesn't really serve any purpose) but I think that may require additional code changes and not just type changes.

**Comment by silviuaavram** [comment]
I reviewed it, @zmdavis , and that's why I brought it up again.

To my mind, there is no reason to update the prop types. If you check the code, we are using an empty object as default, only if window is not available (SSR).

However, this does not need to affect the prop types. The default empty object is internal implementation, it does not relate to user props.

If the user needs to pass a window equivalent via environment, then that equivalent must have have the props marked as required.

So, there are 3 cases:
- window exists, don't do anything.
- window exists or not, but we are in iframe / shadow dom or related land, so we pass the window equivalent via environment. This equivalent must have the required properties, like the iframe example we use on the docsite.
- window does not exist, like in SSR. Don't do anything, downshift uses empty object.

In none of these cases do we need to have an empty object as prop type for environment.

Consequently, you should not pass `environment: {}` as a prop, it's not correct.

**Comment by zmdavis** [comment]
I can work on a contained reproduction when I have time, but the current situation is that prop type warnings are thrown when omitting the `environment` prop altogether (which I think we're in agreement is a reasonable thing to do) and `window` is not available (SSR). So you seem to be suggesting that nothing needs to change, and I respectfully disagree but am happy to take on the burden to prove otherwise.

**Comment by silviuaavram** [comment]
```
  environment: PropTypes.shape({
    addEventListener: PropTypes.func.isRequired,
    removeEventListener: PropTypes.func.isRequired,
    document: PropTypes.shape({
      createElement: PropTypes.func.isRequired,
      getElementById: PropTypes.func.isRequired,
      activeElement: PropTypes.any.isRequired,
      body: PropTypes.any.isRequired,
    }).isRequired,
    Node: PropTypes.func.isRequired,
  }),
```

This is the prop type for `environment`. It is not marked as `isRequired`. I fail to understand why, if I don't provide the `environment` prop to the hook, I would get a prop type warning.

**Comment by zmdavis** [comment]
My understanding is this: the [default prop](https://github.com/downshift-js/downshift/blob/master/src/downshift.js#L112) for `environment` in the SSR case is an empty object, which violates the prop type definition. In this case not passing an `environment` prop is (as far as the component render function is concerned) the same as passing in an empty object. I believe that prop type checking is run on the props passed to the render function and isn't aware of where the props come from (eg, defaultProps or passed in).

```javascript
static defaultProps = {
   ...
   environment: typeof window === 'undefined' ? {} : window,
   ...
}
```

We could _only_ update the propTypes and leave the Typescript types alone, which would be more accurate to the situation but has the downside of the different kinds of types being out of sync with each other.

An alternative fix would be to not set the fallback in `defaultProps` but instead do so within the component itself. That's a slightly more involved change since it's not dealing with just types, but I'm happy to put a PR up for that instead (or in addition for comparison) if you're curious about that solution.

**Comment by silviuaavram** [comment]
Looking into the code, for the hooks we have:
```
function useSelect(userProps = {}) {
  validatePropTypes(userProps, useSelect)
  // Props defaults and destructuring.
  const props = {
    ...defaultProps,
    ...userProps,
  }
```
So the prop type checking is done on the user props, then we use the defaultProps which contain the `{}` default for the `environment`.

In `Downshift` I think this should work by default with the static propTypes and static defaultProps definitions.

So, consequently, I don't understand where the prop types issue comes from if you don't explicitly pass any prop. Maybe just create a sandbox to repro this issue, and we can take it from there. Thanks!

**Comment by silviuaavram** [comment]
I think I'm on to something. Will update here on the progress.

**Comment by jblarriviere** [comment]
Thanks for looking into it. Here is a basic repro with a RN/Expo example, if it helps.
Feel free to let me know if there is more I can do to help

**Comment by silviuaavram** [comment]
https://github.com/downshift-js/downshift/pull/1538 should fix this, along with supporting both react 18 and below 18. It should be available with v8.2.0.

**Comment by jblarriviere** [comment]
@silviuaavram thanks again for taking the time to try and fix this. I've not had a chance to test on SSR yet, but I bumped to 8.2.0 in the repro linked [here](https://github.com/jblarriviere/downshift-repro) and still got the warning

**Comment by silviuaavram** [comment]
OK, so RN has the window object, but without the properties Downshift requires.

```js
["Promise", "__tickleJs", "console", "__jsiExecutorDescription", "nativeModuleProxy", "nativeFlushQueueImmediate", "nativeCallSyncHook", "globalEvalWithSourceUrl", "nativeLoggingHook", "nativePerformanceNow", "_WORKLET_RUNTIME", "_REANIMATED_VERSION_CPP", "__reanimatedModuleProxy", "expo", "ExpoModules", "__BUNDLE_START_TIME__", "__DEV__", "process", "__METRO_GLOBAL_PREFIX__", "__requireCycleIgnorePatterns", "__r", "__d", "__c", "__registerSegment", "$RefreshReg$", "$RefreshSig$", "__accept", "originalConsole", "ErrorUtils", "window", "self", "DOMRect", "DOMRectReadOnly", "__SYSTRACE", "__fbGenNativeModule", "performance", "setTimeout", "clearTimeout", "setInterval", "clearInterval", "requestAnimationFrame", "cancelAnimationFrame", "requestIdleCallback", "cancelIdleCallback", "setImmediate", "clearImmediate", "queueMicrotask", "XMLHttpRequest", "FormData", "fetch", "Headers", "Request", "Response", "WebSocket", "Blob", "File", "FileReader", "URL", "URLSearchParams", "AbortController", "AbortSignal", "alert", "navigator", "__fetchSegment", "__REACT_DEVTOOLS_CONSOLE_FUNCTIONS__", "__blobCollectorProvider", "__ReactRefresh", "__EXAV_setOnAudioSampleReceivedCallback", "__EXAV_onDestroyHostObject", "ExpoModulesCore_CodedError", "__extends", "__assign", "__rest", "__decorate", "__param", "__esDecorate", "__runInitializers", "__propKey", "__setFunctionName", "__metadata", "__awaiter", "__generator", "__exportStar", "__createBinding", "__values", "__read", "__spread", "__spreadArrays", "__spreadArray", "__await", "__asyncGenerator", "__asyncDelegator", "__asyncValues", "__makeTemplateObject", "__importStar", "__importDefault", "__classPrivateFieldGet", "__classPrivateFieldSet", "__classPrivateFieldIn", "__addDisposableResource", "__disposeResources"]
```

**Comment by silviuaavram** [comment]
Will reopen the ticket. I _think_ we should also check isReactNative in the defaultProps assignment.

Sorry to prolong this issue, but apparently it's a bit more complex than it seemed. And the prop type fix we introduced in v8 was actually trying to make users do the right thing: if they supply environment manually as a prop, to make sure it contains the required properties.

The fact that the prop types also check for the default props in a class component was also unexpected.

Will try the isReactNative fix.

**Comment by silviuaavram** [comment]
The fix should be available with 8.2.1

**Comment by jblarriviere** [comment]
Works like a charm, thank you very much !
